commcare_export/cli.py

```diff
@@ -207,19 +218,99 @@ def _get_checkpoint_manager(args):
+            lazy_result = query.eval(env)
+            if lazy_result is not None:
+                # evaluate lazy results
+                for r in lazy_result:
+                    list(r) if r else r
+        except requests.exceptions.RequestException as e:
+            if e.response.status_code == 401:
+                print("\nAuthentication failed. Please check your credentials.")
+                return None
+            else:
+                raise
+        except KeyboardInterrupt:
+            print('\nExport aborted')
+            return None
+
+
+def location_type_map(args):
+    location_type_query = MiniLinq.from_jvalue(builtin_location_types_query.jvalue)
+
+    writer = writers.JValueTableWriter()
+    checkpoint_manager = None
+
+    commcarehq_base_url = commcare_hq_aliases.get(args.commcare_hq,
+                                                  args.commcare_hq)
+    api_client = _get_api_client(args, commcarehq_base_url)
+
+    cm = CheckpointManagerProvider(checkpoint_manager, None, args.start_over)
+    static_env = {
+        'commcarehq_base_url': commcarehq_base_url,
+        'get_checkpoint_manager': cm.get_checkpoint_manager,
+    }
+    env = (
+            BuiltInEnv(static_env)
+            | CommCareHqEnv(api_client, until=None, page_size=args.batch_size)
+            | JsonPathEnv({})
+            | EmitterEnv(writer)
+    )
+
+    evaluate_query(env, location_type_query)
+
+    return builtin_location_types_query.build_dictionary(writer.tables)
+
+
 def main_with_args(args):
     logger.info("CommCare Export Version {}".format(__version__))
     writer = _get_writer(args.output_format, args.output, args.strict_types)
 
-    try:
-        query = _get_query(args, writer)
-    except DataExportException as e:
-        print(e.message)
+    if args.query is None and args.users is False and args.locations is False:
+        print('At least one the following arguments is required: '
+              '--query, --users, --locations')
         return EXIT_STATUS_ERROR
 
-    if not query:
-        print('Query file not found: %s' % args.query)
-        return EXIT_STATUS_ERROR
+    query = None
+    query_tables = []
+    if args.query is not None:
+        try:
+            query = _get_query(args, writer)
+        except DataExportException as e:
+            print(e.message)
+            return EXIT_STATUS_ERROR
+
+        if not query:
+            print('Query file not found: %s' % args.query)
+            return EXIT_STATUS_ERROR
+        else:
+            query_tables = query.emitted_tables()
+
+    if args.users:
+        if builtin_users_query.TABLE_NAME in query_tables:
+            print('--users flag should not be used with a query that writes '
+                  'to the table', builtin_users_query.TABLE_NAME)
+            return EXIT_STATUS_ERROR
+        # Add user data to query
+        user_query = MiniLinq.from_jvalue(builtin_users_query.jvalue)
+        query = MiniLinq.combine(query, user_query)
```

Comment:
I don't think this `combine` function is necessary. It's simpler to do something like this: https://github.com/dimagi/commcare-export/blob/77d4a9e0ad61878bd9aaf35efa00a7922c63b415/commcare_export/excel_query.py#L458-L459

You could also then get rid of `List.concatenate`

Reply:
I tried to make this change but ran into a problem and backed it out.

The problem appears in tests like "test_cli_table_plus_users" where the Excel query emits to two tables, which gave a MiniLinq query like List(List(table1, table2), users). The test failed by not finding the expected output for one of the Excel query produced tables.

So it seems that a nested List query does not behave like a flattened List query. Are they expected to behave the same?

Please let me know if I missed something obvious here.

Reply:
I tried doing this locally and it worked fine and the tests pass:

```python
def get_queries(args, writer):
    query_list = []
    if args.query is not None:
        query = _get_query(args, writer)
        if not query:
            raise MissingQueryFileException(args.query)
        query_list.append(query)

    if args.users:
        # Add user data to query
        query_list.append(builtin_queries.users_query)

    if args.locations:
        # Add location data to query
        query_list.append(builtin_queries.locations_query)

    return List(query_list) if len(query_list) > 1 else query_list[0]
```

Reply:
I made this change, even though it causes two of the new tests to fail locally (test_cli_table_plus_users and test_cli_table_plus_locations). I'm expecting them to fail in Travis CI as well, although all I can see right now is a failed test with an empty log.

Since the tests always failed for me locally, I tracked down the cause. The value of 'lazy_result' in evaluate_query in cli.py was a nested list such as [[RepeatableIterator, None], None] and that function only traversed the top level list, never iterating over the nested RepeatableIterator. Changing that code to perform a recursive traversal fixed the problem for me, but I didn't want to make that change until we agree that there is actually a problem.

Reply:
fine to make that recursively traverse the result.

Reply:
OK, I added the recursive traversal and now only see the integration tests failing in Travis CI, as expected.